zml/module.zig

```diff
@@ -244,26 +243,29 @@ pub const CompilationContext = struct {
 
             var fn_res_values: [out_tensor_count]mlir.Value = undefined;
             self.extractValuesAndTypes(&fn_res, &fn_res_values, fn_res_types, fn_res_shapes, fn_res_donations);
+
             const fn_ret = dialect.func.return_(mlir_ctx, &fn_res_values, loc);
             fn_body.addOperationsRecursive(fn_ret);
         }
 
         const arg_attrs = try arena.alloc(AttributeList, tensor_count);
         @memset(arg_attrs, .{});
 
+        const res_attrs = try arena.alloc(AttributeList, out_tensor_count);
+        @memset(res_attrs, .{});
+
         // Donations attributes only make sense on the main function.
-        if (opts.add_donations_attributes) {
-            self.addDonationsAttributes(arg_attrs, fn_res_donations);
-        }
-        if (opts.sharding) {
-            self.addShardingAttributes(arg_attrs, input_shapes.items);
-        }
+        self.addDonationsAttributes(arg_attrs, fn_res_donations);
```

Comment:
I remember why we had an opts. It was for the function caching. Because inner function can't do donation, so I didn't want to put the donation attributes there. But let's keep it that way for now, we may do it differently next time around.